File: examples/basic_usage.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <thread>

#include "limitless/limiter.hpp"

// Simulated OpenAI response
struct ChatResponse {
    std::string content;
};

// Pretend this calls the real OpenAI API
ChatResponse fake_openai_call(const std::string& prompt) {
    std::this_thread::sleep_for(std::chrono::milliseconds(50)); // simulate latency
    return {"Response to: " + prompt};
}

int main() {
    // Allow 10 requests and 1000 tokens per minute
    limitless::ChatRateLimiter rate_limiter(/*request_limit=*/10,
                                            /*token_limit=*/1000);

    std::vector<std::string> prompts = {
        "What is the capital of France?",
        "Explain quantum computing simply.",
        "Write a haiku about rain.",
    };

    for (const auto& prompt : prompts) {
        std::vector<limitless::Message> messages = {
            {"user", prompt, ""}
        };

        // Blocks here if either limit would be exceeded, then proceeds
        auto guard = rate_limiter.limit("gpt-4", messages, /*max_tokens=*/50);

        auto response = fake_openai_call(prompt);
        std::cout << "Q: " << prompt << "\n"
                  << "A: " << response.content << "\n\n";
    }

    return 0;
}

File: examples/basic_usage.py
"""
Example: using the C++-backed limitless from Python.

Build first:
    pip install pybind11
    python setup.py build_ext --inplace

Then run:
    python examples/basic_usage.py
"""

import openai
import limitless   # <-- this is your compiled C++ extension

# 200 requests/min, 40,000 tokens/min
rate_limiter = limitless.ChatRateLimiter(
    request_limit=200,
    token_limit=40_000,
)

messages = [
    limitless.Message(role="user", content="What is the capital of France?")
]

# --- Option 1: context manager ---
with rate_limiter.limit("gpt-4", messages, max_tokens=100):
    response = openai.chat.completions.create(
        model="gpt-4",
        messages=[{"role": m.role, "content": m.content} for m in messages],
        max_tokens=100,
    )
    print(response.choices[0].message.content)


# --- Option 2: assign guard to variable ---
guard = rate_limiter.limit("gpt-4", messages, max_tokens=100)
response = openai.chat.completions.create(
    model="gpt-4",
    messages=[{"role": m.role, "content": m.content} for m in messages],
    max_tokens=100,
)
print(response.choices[0].message.content)


# --- Option 3: token counting only ---
n = limitless.num_tokens_consumed_by_chat_request(
    model="gpt-4",
    messages=messages,
    max_tokens=100,
)
print(f"Estimated tokens: {n}")